### PR TITLE
Add get_stream_from_external API for XPU backend

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #143849
* #143799
* __->__ #141123
* #141119
* #142347

# Motivation
This PR aims to introduce `torch.xpu.ExternalStream` to be used to wrap SYCL queue created in other libraries to PyTorch.

# Additional Context

cc @gujinghui @EikanWang @fengyuan14

### CODE REVIEW

File: torch/xpu/streams.py [hunk 1]
```diff
@@ -96,6 +96,29 @@ def __repr__(self):
...
+    to facilitate data exchange and multi-library interactions.
+
+    .. note:: This class doesn't manage the queue life-cycle, it is the user
+       responsibility to keep the referenced queue alive while this class is
+       being used. The different SYCL queue pointers will result in distinct
+       ExternalStream objects, even if they reference the same underlying SYCL
+       queue object.
+
+    Args:
+        queue_ptr(int): Integer representation of the `sycl::queue*` value passed externally.
```
Comment (gujinghui) [inline torch/xpu/streams.py:112]: Should we highlight here again. The queue_ptr to a SYCL queue and its copies will generate distinct ExternalStream objects.

Comment (guangyey) [reply torch/xpu/streams.py:111]: I have highlighted it in ..note. which could be rendered. I think it is enough.

File: torch/xpu/streams.py [hunk 2]
```diff
@@ -96,6 +96,28 @@ def __repr__(self):
         return f"torch.xpu.Stream(device={self.device} sycl_queue={self.sycl_queue:#x})"
 
 
+class ExternalStream(Stream):
```
Comment (albanD) [inline torch/xpu/streams.py:99]: Why is this one a cleass and not just a function that creates this very specific Stream object? More or less just a binding to at::xpu::getStreamFromExternal

**Review by albanD — COMMENTED** [review]
Sounds ok in principle, just a quick question if we could keep this more simple

File: torch/xpu/streams.py (hunk 2 above)
Comment (guangyey) [reply torch/xpu/streams.py:99]: Actually, here we follow CUDA's practice. I guess the new class `ExternalStream` could help the user to determine if a stream object is an external stream via `type(stream) == torch.xxx.ExternalStream`. And remind the user to be aware of the limitations of external streams like life cycle issues which is different from the native stream.

If you prefer a new binding function instead of the new class, I could try to re-design it.

Comment (guangyey) [reply torch/xpu/streams.py]: @albanD I add `torch.xxx.get_stream_from_external` API for CUDA and XPU. I don't think each kind of accelerator has the concept of the external stream. For mps, it seems like only have a stream. So I don't add this API to torch.accelerator namespace. What do you think about it?

File: torch/csrc/xpu/Module.cpp [hunk 3]
```diff
@@ -390,6 +390,17 @@ static void initXpuMethodBindings(PyObject* module) {
       "torch.xpu.mem_get_info requires PyTorch to be built with SYCL compiler version 2025.0.0 or newer.");
 #endif
   });
+  m.def(
```
Comment (albanD) [inline torch/csrc/xpu/Module.cpp:393]: Thanks for using pybind here for easier to read python binding! :)

File: torch/csrc/xpu/Stream.cpp [hunk 4]
```diff
@@ -22,20 +22,27 @@ static PyObject* THXPStream_pynew(
...
+      "stream_id",
+      "device_index",
+      "device_type",
+      "queue_ptr",
+      nullptr};
   if (!PyArg_ParseTupleAndKeywords(
           args,
           kwargs,
-          "|iLLL",
+          "|iLLLK",
```
Comment (albanD) [inline torch/csrc/xpu/Stream.cpp:38]: Why did you add this to the base Stream constructor? We don't actually use it here?

**Review by albanD — COMMENTED** [review]
Thanks!
Just a quick test to make sure it behaves as expected and why the Stream object constructor was changed!

File: torch/csrc/xpu/Stream.cpp (hunk 4 above)
Comment (guangyey) [reply torch/csrc/xpu/Stream.cpp]: I originally wanted to align with the constructor of CUDA's base Stream. It is okay to remove it. And add it back if really needed in future.

**Comment by guangyey** [comment]
> Thanks! Just a quick test to make sure it behaves as expected and why the Stream object constructor was changed!

It is hard to add a quick test to validate this API. Unlike [cudart](https://github.com/pytorch/pytorch/blob/4bacfd6e117f1e77dbce9444fc3ef3d95c5f2b89/torch/csrc/cuda/shared/cudart.cpp#L89) in PyTorch, XPU couldn't create an external sycl queue by PyTorch since it doesn't have the wrapper of sycl runtime. I have to test this API on my local machine.

**Review by albanD — APPROVED** [review]
Ok!

**Review by EikanWang — APPROVED** [review]
